src/tape/ensemble.py

```diff
@@ -807,6 +807,130 @@ def s2n_inter_quartile_range(flux, err):
         else:
             return batch
 
+    def from_pandas(
+        self,
+        source_frame,
+        object_frame=None,
+        column_mapper=None,
+        sync_tables=True,
+        npartitions=None,
+        partition_size=None,
+        **kwargs,
+    ):
+        """Read in Pandas dataframe(s) into an ensemble object
+
+        Parameters
+        ----------
+        source_frame: 'pandas.Dataframe'
+            A Dask dataframe that contains source information to be read into the ensemble
+        object_frame: 'pandas.Dataframe', optional
+            If not specified, the object frame is generated from the source frame
+        column_mapper: 'ColumnMapper' object
+            If provided, the ColumnMapper is used to populate relevant column
+            information mapped from the input dataset.
+        sync_tables: 'bool', optional
+            In the case where an `object_frame`is provided, determines whether an
+            initial sync is performed between the object and source tables. If
+            not performed, dynamic information like the number of observations
+            may be out of date until a sync is performed internally.
+        npartitions: `int`, optional
+            If specified, attempts to repartition the ensemble to the specified
+            number of partitions
+        partition_size: `int`, optional
+            If specified, attempts to repartition the ensemble to partitions
+            of size `partition_size`.
+
+        Returns
+        ----------
+        ensemble: `tape.ensemble.Ensemble`
+            The ensemble object with the Dask dataframe data loaded.
+        """
+        # Construct Dask DataFrames of the source and object tables
+        source = dd.from_pandas(source_frame, npartitions=npartitions)
+        object = None if object_frame is None else dd.from_pandas(object_frame, npartitions=npartitions)
```

Comment:
These are good questions!

1. If I'm understanding correctly, in cases where we can fit our data in a pandas dataframe in RAM, we honestly probably don't gain anything from using Dask (see [Dask best practices](https://docs.dask.org/en/stable/dataframe-best-practices.html#use-pandas)). The `from_pandas` and `from_source_dict` cases are more for allowing a user to play around with smaller data sets, but I don't think they'll be very performant. I think if we want to support those use cases more, we'll probably want the `source` and `object` frames to be in pandas rather than Dask. @dougbrn might have opinions here when he gets back.

2. A call to `dask.dataframe.repartitions` where `npartitions` is the frame's current number of partitions is a no-op (see [here](https://github.com/dask/dask/blob/main/dask/dataframe/core.py#L7854C38-L7854C38)). So my thoughts were to go ahead and specify the number of partitions at frame creation in the hopes that we might save the cost of partitioning twice. Calling repartition in `from_dask_dataframe` of course just gives the user an easy way to repartition their frame if needed.

However in `from_parquet` we don't specify the number of partitions in the object table. So removing specifying npartitions for the object table here to conform to that behavior.

3. Yes having a partitioning scheme to avoid/limit this is important. Note that the source frame's index is based on the object ID, so dask's partitioning scheme should be trying to group together source data by object. But it might turn out we need to add some more intelligent ways to prevent this. I believe you are correct about nodes and partitions.

4. Yeah, I'm worried about this as well. The fact that both frames should be sorted by the same object ID index should allow Dask to try to avoid/limit cases like these (see about sorted joins [here](https://docs.dask.org/en/stable/dataframe-joins.html#sorted-joins)), but I think we'll need to see what the costs are in practice. I've modified `from_dask_dataframe` to ensure the object table is always indexed by the object ID regardless of how the user created the object frame.

And yeah if you feel like you want to followup over Slack rather than here, feel free to reach out and I can then summarize the discussion here :)

Reply:
Thank you!